### DISCUSSION TITLE
Problem Adding PDF content in AIR application with 64 bit Adobe Acrobat Reader

### DISCUSSION THREAD

**Discussion by concord232** [airsdk/Adobe-Runtime-Support#1807] [Q&A]
For our desktop application we have been using PDF render capability (https://help.adobe.com/en_US/as3/dev/WS5b3ccc516d4fbf351e63e3d118666ade46-7eb4.html) . It works fine with 32bit Adobe Acrobat reader. 

But recently Adobe has been rolling out 64 bit versions and PDF render does not work with the 64 bit version. When checking for rendering capability using the following code 

> if(HTMLLoader.pdfCapability == HTMLPDFCapability.STATUS_OK)  
> { 
>     trace("PDF content can be displayed"); 
> } 
> else  
> { 
>     trace("PDF cannot be displayed. Error code:", HTMLLoader.pdfCapability); 
> } 
> 

we have been getting error HTMLPDFCapability.ERROR_INSTALLED_READER_TOO_OLD which is quite strange. The Adobe reader installed is the latest version and definitely > 8.1, just that it is 64 bit version. Now if I uninstall the 64 bit version and install back the 32 bit version it works fine.

Tried using both AIR 32.0 and the latest AIR 33.1 (Runtime and SDK)

Any suggestions pls

**Comment by ajwfrost** [comment]
If there's no 32-bit support for PDF then I would suggest running your application as 64-bit? Windows can't load a library of one architecture into an application running in the other..

We can look at the error code, to see if there's a better value that can be returned..

thanks

**Reply by concord232** [reply]
Thanks for replying, really appreciate.

We tried building the application as 64bit with AIR SDK v33.1 the application failed to launch with error "Initial content could not be loaded for this application. Try reinstalling or contact the published for assitance" 

So we tried building 64bit application with AIR SDK v32 this time the application was able to launch, but it crashed when tried to render the PDF. We tried debugging even debugger crashed.

(Same code works well when building for 32bit using both AIR SDK v33.1 and v32, the issue popups up only when we try to build for 64bit version)

So now we have no way to run the application using 64bit Adobe Acrobat. 

Based on similar post -> https://github.com/airsdk/Adobe-Runtime-Support/discussions/1411 I could see that you will be contacting the Adobe. Any updates please in that regards? can we see any release with the fix in future?

**Reply by rogerwilliam1990** [reply]
We are also facing the same issue. were you able to find any solution?

**Reply by ajwfrost** [reply]
Hi

I'd not got any feedback from Adobe on this, but will chase up .. but I suspect it's just that the 32-bit support has been dropped, hence no longer working with AIR in 32-bit mode. In our next major release, we're planning to shift everything over to use 64-bit on Windows by default which will hopefully help..!

In terms of the applications not starting, with an error about the initial content, this appears to be caused by the removal of the internal webkit library meaning that the ability of AIR to have initial content being an HTML file no longer works. We've been looking at re-enabling this, and it looks like this should be feasible, although the interactions between the HTML content and the AIR engine may be impacted by the change here, so there may need to be some work to change this. (Or, take an older AIR SDK and build your application with captive runtime i.e. as a bundle).

The challenge then is to see whether an application can open the 64-bit reader plug-in when it's running in 64-bit mode itself .. and it seems to, for us...? We're using a SWF file with the below code in it (taken off a sample website):
```
			var strURL : String = "https://www.w3docs.com/uploads/media/default/0001/01/540cb75550adf33f281f29132dddd14fded85bfc.pdf";
			var pdf : HTMLLoader = new HTMLLoader();
			pdf.height = 800;
			pdf.width = 600;
			var strHTML : String = "<!DOCTYPE html>\n" +
				"<html>\n" +
				"  <head>\n" +
				"    <title>Title of the document</title>\n" +
				"  </head>\n" +
				"  <body>\n" +
				"    <h1>PDF Example with iframe</h1>\n" +
				"    <iframe src=\"" + strURL + "\" width=\"100%\" height=\"500px\">\n" +
				"    </iframe>\n" +
				"  </body>\n" +
				"</html>";
			pdf.loadString(strHTML);	
			addChild(pdf);		
```
although it should also work if you used StageWebView directly, with resize handlers to update the viewport as required...

Are you able to confirm whether you can view a PDF when using this sort of set-up?

thanks

**Reply by concord232** [reply]
Hi @ajwfrost,

Thanks for replying

Regarding Application not starting, Same application works with v33.1 SDK when packaged as native which is 32 bit. But when we try packaging using bundle option it has this problem of not able to launch.

We are using HTML and not SWF. We tried to build 64 bit application with bundle option using v32 SDK, the app launched but it crashed while trying to render the pdf content. We used similar code as above for rendering the PDF.

We hope with next SDK release with the fix once the loading HTML issue is solved for 64 bit application packaged with bundle, we are able to render the PDF.
When can we expect the release ? Any rough timeline would really help. Thanks!

**Reply by concord232** [reply]
> We are also facing the same issue. were you able to find any solution?

Hi @rogerwilliam1990 , no solution yet. Thanks!

**Reply by ajwfrost** [reply]
@concord232 so trying to confirm the current situation:

1) if you package an application using AIR SDK 33.1 and the "native" option, then it will be a 32-bit installer but without the runtime -> if you install it, and it works, are you able to confirm what version of the AIR shared runtime is on your machine (via Add/Remove Programs, it should have AIR in the list and can show you the version details..)

2) if you package it using AIR SDK 33.1 and the "bundle" option, then it will be 32-bit with a bundled runtime -> currently this doesn't load content that are HTML files.

3) if you package it using AIR SDK 32.0 and the "bundle" option, then if you use 64-bit then it's crashing .. not sure why that would be, but it may be just some incompatibility between the new Acrobat plug-in and the old WebKit browser engine.

We have made an update to the AIR SDK 33.1 that re-enables the idea of loading in an HTML file. Trying it out on Windows though, where the HTML just linked to a PDF, it caused it to trigger the download of the PDF file rather than displaying it inline. To display it inline we needed our HTML to be using that iframe mechanism...

We've been looking at another application which is HTML-based which isn't working with this change, because it's trying to do additional interactions between the HTML/JavaScript and the AIR runtime. I'm not sure how much of the original integration we can manage to get working again, but we're looking into this still. We'll probably go with a "step 1" release where the basic functionality is available, estimated release date would be at the start of May. I have some concerns that the crash you're seeing with AIR 32 may then happen again with what we're doing though .. so it could be handy to get a pre-release over to you so you can check against your particular machine configuration (and Acrobat installation etc..)..


thanks

**Reply by concord232** [reply]
Hi @ajwfrost 

Thanks for the details 

I have complied and listed the different observations below

**The objective is to make application work with 64bit Adobe Reader.** Kindly do note we also have the use of PDF-HTML communication (https://help.adobe.com/en_US/air/html/dev/WS5b3ccc516d4fbf351e63e3d118666ade46-7eb4.html#WS5b3ccc516d4fbf351e63e3d118676a58f8-7ffa)

Following are the behavior observed

1.  v32 SDK, Native with v32 runtime , 32 bit  application====> Loads HTML, Renders PDF but only works with 32 bit Adobe reader, for 64 bit Adobe reader it fails to render and we would see white screen.
2. v32 SDK, Bundle, 64 bit application====> Loads HTML, but crashes while rendering PDF when used with 64 bit Adobe reader.
3. v33.1 SDK, Native with v33.1 runtime, 32 bit application===>  Loads HTML, Renders PDF but only works with 32 bit Adobe reader, for 64 bit Adobe reader it fails to render and we would see white screen.
4. v33.1 SDK, Bundle, 64 bit application ====> Fails to load HTML

We are urgently looking for any solution which would enable us rendering PDF using 64 bit Adobe Reader.

Thank you for your support, would be happy to help in any way to speed this process including testing the functionality.

**Reply by ajwfrost** [reply]
Thanks: so with your fourth case there, you should have the bundle folder which then has a sub-folder "Adobe AIR\Versions\1.0". In there is the "Adobe AIR.dll" file ... can you try replacing that with the one from the below message, and see whether that then loads the HTML?
https://transfer.harman.com/message/5XKnivyVZg256KdiSxLGub

thanks

**Reply by concord232** [reply]
@ajwfrost Thanks! 

I tried the same. This time I didn't see the pop message with the error.

But application failed to load the menu and to show the HTML page with the image

**Reply by ajwfrost** [reply]
Hmm... would there be any possibility that we can get hold of your application and try it? or can you create a simpler bit of html content that demonstrates the issue?

The other thing to check: it will be using the OS implementation i.e. MSHTML (IE-based) webbrowser control, unless you have the Edge "WebView2" component installed. So it depends a little on how your HTML is written, whether it works in IE vs Edge/Chrome, etc...

---

Part of me is thinking: we could actually just re-enable the embedded webkit for Windows apps, as the only real problem we have is with M1/apple silicon on the new macOS platforms. But it would have been nice to get rid of this old webkit library completely.....

**Reply by concord232** [reply]
Hi @ajwfrost I quickly created a similar sample application. 


[Sample App 2.zip](https://github.com/airsdk/Adobe-Runtime-Support/files/8869623/Sample.App.2.zip)

I have organized the source/ adt  command used/The Package created in the zip file. More info about the folder and application in the readme file. 

I hope this is helpful. Kindly help with the SDK.

Thanks!

**Reply by concord232** [reply]
Hi @ajwfrost I modified the above shared application to include PDF-JS Communication. Kindly read the readme file for details

[32bit Bundle_Build Using AIR SDK 33.1.1.743.zip](https://github.com/airsdk/Adobe-Runtime-Support/files/9367568/32bit.Bundle_Build.Using.AIR.SDK.33.1.1.743.zip)
[32bit Native_Build Using AIR SDK 32.0.zip](https://github.com/airsdk/Adobe-Runtime-Support/files/9367570/32bit.Native_Build.Using.AIR.SDK.32.0.zip)
[64bit Bundle_Build Using AIR SDK 33.1.1.743.zip](https://github.com/airsdk/Adobe-Runtime-Support/files/9367571/64bit.Bundle_Build.Using.AIR.SDK.33.1.1.743.zip)

**Reply by concord232** [reply]
Hi @ajwfrost this is the application with the WebKit dll replaced

[64bit Bundle_Build using AIR SDK 33.1.1.743_With WebKit Dll replaced in SDK.zip](https://github.com/airsdk/Adobe-Runtime-Support/files/9523861/64bit.Bundle_Build.using.AIR.SDK.33.1.1.743_With.WebKit.Dll.replaced.in.SDK.zip)
